engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java

```diff
@@ -0,0 +1,396 @@
+  Object readArrayElement(long index)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    throw CompilerDirectives.shouldNotReachHere();
+  }
+
+  @ExportMessage
+  final void writeArrayElement(long index, Object value) throws UnsupportedMessageException {
+    throw UnsupportedMessageException.create();
+  }
+
+  @ExportMessage
+  boolean isArrayElementInsertable(long index) {
+    return false;
+  }
+
+  @ExportMessage
+  boolean isArrayElementRemovable(long index) {
+    return false;
+  }
+
+  @ExportMessage
+  final void removeArrayElement(long index) throws UnsupportedMessageException {
+    throw UnsupportedMessageException.create();
+  }
+
+  @ExportMessage
+  @CompilerDirectives.TruffleBoundary
+  String toDisplayString(boolean allowSideEffects) {
+    final InteropLibrary iop = InteropLibrary.getUncached();
+    return DisplayArrayUtils.toDisplayString(this, allowSideEffects, iop);
+  }
+
+  @ExportMessage
+  Type getMetaObject(@CachedLibrary("this") InteropLibrary thisLib) {
+    return EnsoContext.get(thisLib).getBuiltins().vector();
+  }
+
+  @ExportMessage
+  boolean hasMetaObject() {
+    return true;
+  }
+
+  //
+  // methods for TypesLibrary
+  //
+
+  @ExportMessage
+  boolean hasType() {
+    return true;
+  }
+
+  @ExportMessage
+  Type getType(@CachedLibrary("this") TypesLibrary thisLib, @Cached("1") int ignore) {
+    return EnsoContext.get(thisLib).getBuiltins().vector();
+  }
+
+  //
+  // helper methods
+  //
+
+  @Override
+  @CompilerDirectives.TruffleBoundary
+  public String toString() {
+    return toDisplayString(false);
+  }
+
+  static Vector fromInteropArray(Object arr) {
+    return new Generic(arr);
+  }
+
+  static Vector fromLongArray(long[] arr) {
+    return new Long(arr);
+  }
+
+  static Vector fromDoubleArray(double[] arr) {
+    return new Double(arr);
+  }
+
+  static Object fromEnsoOnlyArray(Object[] arr) {
+    return new EnsoOnly(arr);
+  }
+
+  @ExportLibrary(InteropLibrary.class)
+  @ExportLibrary(WarningsLibrary.class)
+  static final class EnsoOnly extends Vector {
+    private final Object[] storage;
+
+    private EnsoOnly(Object[] storage) {
+      this.storage = storage;
+    }
+
+    //
+    // messages for the InteropLibrary
+    //
+
+    @ExportMessage
+    long getArraySize() {
+      return storage.length;
+    }
+
+    @ExportMessage
+    EnsoObject readArrayElement(long index) throws InvalidArrayIndexException {
+      try {
+        return (EnsoObject) storage[Math.toIntExact(index)];
+      } catch (ArithmeticException | IndexOutOfBoundsException ex) {
+        throw InvalidArrayIndexException.create(index);
+      }
+    }
+
+    @ExportMessage
+    boolean isArrayElementReadable(long index) {
+      var size = storage.length;
+      return index < size && index >= 0;
+    }
+
+    @ExportMessage
+    boolean hasWarnings() {
+      return false;
+    }
+
+    @ExportMessage
+    Warning[] getWarnings(Node location) throws UnsupportedMessageException {
```

Comment:
@hubertp - I am not sure about this semantics - if the `Vector` has no warnings, is it OK to return `new Warning[0]` regardless of some items having warnings?

Reply:
No, that's wrong. Should collect warnings from all elements.

Reply:
Yup. That's what I had to do in 3c439ae22914a7a028a8d6f9882a86e0a9a76dd5 - as soon as I find a warning I return back to previous `new Vector.Generic(new Array(storage))`. Then the `Array` computes the warning from all elements and the `Vector.Generic` just delegates to it. That's suboptimal from performance perspective, but it is correct from the semantic one.